### PR TITLE
AIP-72: Handle External update TI state in Supervisor

### PR DESCRIPTION
**Pull request by kaxil**
- Updated logic to handle externally updated TI state in Supervisor. This states could have been externally changed via UI, CLI, API etc
- Replaced `FASTEST_HEARTBEAT_INTERVAL` and `SLOWEST_HEARTBEAT_INTERVAL` with `MIN_HEARTBEAT_INTERVAL` and `HEARTBEAT_THRESHOLD` for better clarity

This is part of my efforts to port LocalTaskJob tests to Supervisor: https://github.com/apache/airflow/issues/44356.

This ports over `TestLocalTaskJob.test_mark_{success,failure}_no_kill`.

This PR also allows retrying heartbeats:

- Added `_last_successful_heartbeat` and `_last_heartbeat_attempt` for better separation of tracking successful heartbeats and retries.
- `MIN_HEARTBEAT_INTERVAL` is now respected between heartbeat attempts, even after failures.
-  The num of retries is configurable via `MAX_FAILED_HEARTBEATS`

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py [hunk 1]
```diff
@@ -412,10 +414,6 @@ def _monitor_subprocess(self):
         - Sends heartbeats to the client to keep the task alive
         - Checks if the subprocess has exited
         """
-        # Until we have a selector for the process, don't poll for more than 10s, just in case it exists but
-        # doesn't produce any output
-        max_poll_interval = 10
```
Comment (kaxil) [inline task_sdk/src/airflow/sdk/execution_time/supervisor.py:432]: I just replaced this too with `MIN_HEARTBEAT_INTERVAL`. I do need to add `overtime` feature in the Supervisor -- will do in separate PR.

But let me know if you disagree with this @ashb

Comment (ashb) [reply task_sdk/src/airflow/sdk/execution_time/supervisor.py:432]: This was previously to make sure that we call `self._process.wait(timeout=0)` at most every 10s (since we spend most of our time waiting on Readable events on the sockets.)

How ever that wasn't needed -- as the socket being closed, which happens when the subprocess exits even if it produces no output, shows up as a readable event (where upon reading gives the EOF) so that returns us from the select already.

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py [hunk 2]
```diff
@@ -451,10 +449,38 @@ def _check_subprocess_exit(self):
 
     def _send_heartbeat_if_needed(self):
         """Send a heartbeat to the client if heartbeat interval has passed."""
-        if time.monotonic() - self._last_heartbeat >= FASTEST_HEARTBEAT_INTERVAL:
+        # If the process has exited, we don't need to send any more heartbeats
+        if self._exit_code is not None:
+            return
+
+        if time.monotonic() - self._last_heartbeat >= MIN_HEARTBEAT_INTERVAL:
```
Comment (ashb) [inline task_sdk/src/airflow/sdk/execution_time/supervisor.py]: Nit: could we refactor this method to return early if a heartbeat isn't needed? if `if` blocks get long I generally perfer to follow the rule of "less indented code is simpler to understand".

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py [hunk 3]
```diff
@@ -451,10 +449,38 @@ def _check_subprocess_exit(self):
 
     def _send_heartbeat_if_needed(self):
         """Send a heartbeat to the client if heartbeat interval has passed."""
-        if time.monotonic() - self._last_heartbeat >= FASTEST_HEARTBEAT_INTERVAL:
+        # If the process has exited, we don't need to send any more heartbeats
+        if self._exit_code is not None:
+            return
```
Comment (ashb) [inline task_sdk/src/airflow/sdk/execution_time/supervisor.py]: This isn't strictly true. The process has exit, but if we are entering this function then the send final request to the server _probably_ hasn't happened yet as it is after the return of `_monitor_subprocess`, in which case as long as the supervisor is alive we should continue to heartbeat. (i.e. the server side state of this TI is running still, so we should heartbeat it still.)

The exception (right now) is deferred state which we will send earlier (or we will do once Amogh's PR is merged)

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py [hunk 4]
```diff
@@ -451,10 +449,38 @@ def _check_subprocess_exit(self):
...
+
+        if time.monotonic() - self._last_heartbeat >= MIN_HEARTBEAT_INTERVAL:
             try:
                 self.client.task_instances.heartbeat(self.ti_id, pid=self._process.pid)
                 self._last_heartbeat = time.monotonic()
+            except ServerResponseError as e:
+                # TODO: Should we instead check httpx.HTTPStatusError?
```
Comment (ashb) [inline task_sdk/src/airflow/sdk/execution_time/supervisor.py]: HTTPStatusError will happen when we get a 4xx/5xx that we can't parse.

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py [hunk 5]
```diff
@@ -451,10 +449,38 @@ def _check_subprocess_exit(self):
...
+                            current_pid=e.detail.get("current_pid"),
+                        )
+                elif e.response.status_code == 404:
+                    log.error("Task Instance not found")
+                else:
+                    # TODO: Handle other errors
+                    raise
+
+                # If heartbeating raises an error, kill the subprocess
+                self.kill(signal.SIGTERM)
```
Comment (ashb) [inline task_sdk/src/airflow/sdk/execution_time/supervisor.py]: We don't want to kill the task on the first failure, we want to deal with temporary network glitches and blips and retry (next time around the loop, not a tenancity retry here I don't think) before killing the task proc

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py [hunk 6]
```diff
@@ -451,10 +449,38 @@ def _check_subprocess_exit(self):
...
+                            "message", "Task is no longer in the running state and task should terminate"
+                        )
+                        log.error(error_msg, current_state=e.detail.get("current_state"))
+                    elif reason == "running_elsewhere":
+                        error_msg = e.detail.get("message", "Task is already running elsewhere")
+                        log.error(
+                            error_msg,
+                            current_hostname=e.detail.get("current_hostname"),
+                            current_pid=e.detail.get("current_pid"),
+                        )
```
Comment (ashb) [inline task_sdk/src/airflow/sdk/execution_time/supervisor.py]: Could we simplify this and do `log.error("Server indicated we shouldn't be running anymore", detail=e.detail)`? How does structlog cope with that?

Comment (ashb) [reply task_sdk/src/airflow/sdk/execution_time/supervisor.py]:
```pycon
>>> structlog.get_logger().info("hi", detail={"reason": "not_running"})
2024-11-27 11:57:02 [info     ] hi                             detail={'reason': 'not_running'}
```

Seems okay?

File: task_sdk/tests/execution_time/test_supervisor.py [hunk 7]
```diff
@@ -258,6 +261,55 @@ def test_run_simple_dag(self, test_dags_dir, captured_logs, time_machine):
             "timestamp": "2024-11-07T12:34:56.078901Z",
         } in captured_logs
 
+    def test_state_conflict_on_heartbeat(self, captured_logs, monkeypatch, mocker):
+        """
+        Test that ensures that the Supervisor does not cause the task to fail if the Task Instance is no longer
+        in the running state.
+        """
+        structlog.configure(wrapper_class=structlog.make_filtering_bound_logger(logging.ERROR))
```
Comment (ashb) [inline task_sdk/tests/execution_time/test_supervisor.py]: We should probably make this an autouse fixture in this class since we have it in every test.

File: task_sdk/tests/execution_time/test_supervisor.py [hunk 8]
```diff
@@ -258,6 +261,55 @@ def test_run_simple_dag(self, test_dags_dir, captured_logs, time_machine):
...
+
+        def handle_request(request: httpx.Request) -> httpx.Response:
+            if request.url.path == f"/task-instances/{ti_id}/heartbeat":
+                return httpx.Response(
+                    409,
+                    json={
+                        "reason": "not_running",
+                        "message": "TI is no longer in the running state and task should terminate",
+                        "current_state": "success",
+                    },
```
Comment (ashb) [inline task_sdk/tests/execution_time/test_supervisor.py]: Is it worth having one request succeed, and the next one fail?

File: task_sdk/tests/execution_time/test_supervisor.py (hunk 7 above)
Comment (kaxil) [reply task_sdk/tests/execution_time/test_supervisor.py]: moved it to `captured_logs` fixture fixed in https://github.com/apache/airflow/pull/44406/commits/d57bd63398b9f667f8f179008b05f98a2c74e810

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py (hunk 6 above)
Comment (kaxil) [reply task_sdk/src/airflow/sdk/execution_time/supervisor.py]: good call - resolved in https://github.com/apache/airflow/pull/44406/commits/d57bd63398b9f667f8f179008b05f98a2c74e810

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py (hunk 2 above)
Comment (kaxil) [reply task_sdk/src/airflow/sdk/execution_time/supervisor.py]: Done: https://github.com/apache/airflow/pull/44406/commits/d57bd63398b9f667f8f179008b05f98a2c74e810

File: task_sdk/tests/execution_time/test_supervisor.py (hunk 8 above)
Comment (kaxil) [reply task_sdk/tests/execution_time/test_supervisor.py]: Yeah, done in https://github.com/apache/airflow/pull/44406/commits/d57bd63398b9f667f8f179008b05f98a2c74e810

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py (hunk 5 above)
Comment (kaxil) [reply task_sdk/src/airflow/sdk/execution_time/supervisor.py]: added retry logic in https://github.com/apache/airflow/pull/44406/commits/d71cf35287fe01c3aa50bf2e0763c25ec44be061

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py (hunk 3 above)
Comment (kaxil) [reply task_sdk/src/airflow/sdk/execution_time/supervisor.py]: Removed

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py [hunk 9]
```diff
@@ -466,14 +471,52 @@ def _check_subprocess_exit(self):
...
+            ti_id=self.ti_id,
+            max_retries=MAX_FAILED_HEARTBEATS,
+            exc_info=True,
+        )
+        # If we've failed to heartbeat too many times, kill the process
+        if self.failed_heartbeats >= MAX_FAILED_HEARTBEATS:
+            log.error(
+                "Too many failed heartbeats; terminating process", failed_heartbeats=self.failed_heartbeats
+            )
+            self.kill(signal.SIGTERM)
```
Comment (kaxil) [inline task_sdk/src/airflow/sdk/execution_time/supervisor.py:539]: In a follow-up PR, I will update logic in `self.kill` to use  `SIGINT` and do `SIGTERM` after, and `SIGKILL` if the process is still alive.

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py (hunk 1 above)
Comment (kaxil) [reply task_sdk/src/airflow/sdk/execution_time/supervisor.py:432]: Makes sense, I added some more comments to that code for my understanding

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py [hunk 10]
```diff
@@ -265,7 +270,13 @@ class WatchedSubprocess:
     _terminal_state: str | None = None
     _final_state: str | None = None
 
-    _last_heartbeat: float = 0
+    _last_successful_heartbeat: float = 0
+    _last_heartbeat_attempt: float = 0
```
Comment (ashb) [inline task_sdk/src/airflow/sdk/execution_time/supervisor.py]:
```suggestion
    _last_successful_heartbeat: float = attrs.Field(default=0, init=False)
    _last_heartbeat_attempt: float = attrs.Field(default=0, init=False)
```

to make it not show up as args in init

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py [hunk 11]
```diff
@@ -265,7 +270,13 @@ class WatchedSubprocess:
...
+
+    # After the failure of a heartbeat, we'll increment this counter. If it reaches `MAX_FAILED_HEARTBEATS`, we
+    # will kill the process. This is to handle temporary network issues etc. ensuring that the process
+    # does not hang around forever.
+    failed_heartbeats: int = 0
```
Comment (ashb) [inline task_sdk/src/airflow/sdk/execution_time/supervisor.py]:
```suggestion
    failed_heartbeats: int = attrs.Field(default=0, init=False)
```

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py [hunk 12]
```diff
@@ -423,38 +434,54 @@ def _monitor_subprocess(self):
 
         This function:
 
-        - Polls the subprocess for output
-        - Sends heartbeats to the client to keep the task alive
-        - Checks if the subprocess has exited
+        - Waits for activity on file objects (e.g., subprocess stdout, stderr) using the selector.
```
Comment (ashb) [inline task_sdk/src/airflow/sdk/execution_time/supervisor.py]: Logs and requests socket too

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py [hunk 13]
```diff
@@ -423,38 +434,54 @@ def _monitor_subprocess(self):
...
+            socket_handler = key.data
+
+            # Example of handler behavior:
+            # If the subprocess writes "Hello, World!" to stdout:
+            # - `socket_handler` reads and processes the message.
+            # - If EOF is reached, the handler returns False to signal no more reads are expected.
+            need_more = socket_handler(key.fileobj)
+
+            # If the handler signals that the file object is no longer needed (EOF, closed, etc.)
+            # unregister it from the selector to stop monitoring and close it cleanly.
```
Comment (ashb) [inline task_sdk/src/airflow/sdk/execution_time/supervisor.py]:
```suggestion
            # unregister it from the selector to stop monitoring; `wait()` blocks until all selectors
            # are removed
```

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py [hunk 14]
```diff
@@ -466,14 +493,48 @@ def _check_subprocess_exit(self):
...
+        self._last_heartbeat_attempt = time.monotonic()
+        try:
+            self.client.task_instances.heartbeat(self.ti_id, pid=self._process.pid)
+            # Update the last heartbeat time on success
+            self._last_successful_heartbeat = time.monotonic()
+
+            # Reset the counter on success
+            self.failed_heartbeats = 0
+        except ServerResponseError as e:
+            if e.response.status_code in {409, 404}:
```
Comment (ashb) [inline task_sdk/src/airflow/sdk/execution_time/supervisor.py]: Should we use the [http.HTTPStatus enum](https://docs.python.org/3/library/http.html#http.HTTPStatus) here?

```suggestion
            if e.response.status_code in {HTTPStatus.CONFILCT, HTTPStatus.NOT_FOUND}:
```

File: task_sdk/tests/execution_time/test_supervisor.py [hunk 15]
```diff
@@ -299,6 +289,146 @@ def handle_request(request: httpx.Request) -> httpx.Response:
...
+            return httpx.Response(status_code=204)
+
+        proc = WatchedSubprocess.start(
+            path=os.devnull,
+            ti=TaskInstance(id=ti_id, task_id="b", dag_id="c", run_id="d", try_number=1),
+            client=make_client(transport=httpx.MockTransport(handle_request)),
+            target=subprocess_main,
+        )
+
+        # Wait for the subprocess to finish
```
Comment (ashb) [inline task_sdk/tests/execution_time/test_supervisor.py]:
```suggestion
        # Wait for the subprocess to finish -- it should have been termianted
```

**Review by ashb — APPROVED** [review]
Few nits, LGTM!

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py (hunk 9 above)
Comment (kaxil) [reply task_sdk/src/airflow/sdk/execution_time/supervisor.py:539]: https://github.com/apache/airflow/pull/44465
